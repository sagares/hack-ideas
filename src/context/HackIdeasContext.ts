import { createContext } from "react";
import { IdeasArray } from "../HackIdea.types";

type Employee = {
  empId: string;
};

type HackIdeasContextType = {
  currentUser: Employee;
  logout: () => void;
  setCurrentUser: (employee: Employee) => void;
  ideas: IdeasArray;
  setIdeas: (ideas: IdeasArray) => void;
};

const HackIdeasContext = createContext<Partial<HackIdeasContextType>>({});

export { HackIdeasContext };
