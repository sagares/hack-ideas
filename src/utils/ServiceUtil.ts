// Import the functions you need from the SDKs you need
import { initializeApp } from "firebase/app";
import { getDatabase, ref, onValue, set } from "firebase/database";
import {  Idea } from "../HackIdea.types";

const firebaseConfig = {
  apiKey: "AIzaSyC0RFn0RNnX8EoPHpAu-lf4EvN08aPKMCg",
  authDomain: "hack-ideas-c7cd7.firebaseapp.com",
  databaseURL: "https://hack-ideas-c7cd7-default-rtdb.firebaseio.com",
  projectId: "hack-ideas-c7cd7",
  storageBucket: "hack-ideas-c7cd7.appspot.com",
  messagingSenderId: "197788164361",
  appId: "1:197788164361:web:36a94bf614b4b5c830f3d8",
  measurementId: "G-660ZLLJJEB",
};

// Initialize Firebase
const app = initializeApp(firebaseConfig);
const database = getDatabase();

const registerEmployee = (empId: string) => {
  set(ref(database, "employees/" + empId), {
    empId,
  });
};

const getEmployeeById = (empId: string, callback: Function) => {
  let empRef = ref(database, "employees/" + empId);
  return onValue(empRef, (snapshot) => {
    callback(snapshot.val());
  });
};

const getEmployeeData = (callback: Function) => {
  let empRef = ref(database, "employees/");
  return onValue(empRef, (snapshot) => {
    callback(snapshot.val());
  });
};

const getIdeas = (callback: Function) => {
  let empRef = ref(database, "ideas/");
  return onValue(empRef, (snapshot) => {
    callback(snapshot.val());
  });
};

const registerIdea = (idea: Idea) => {
  const id = toCamelCase(idea.name);
  set(ref(database, "ideas/" + id), {
    ...idea,
    id,
    upvotes: [],
  });
};

const updateIdea = (ideaId: string, updatedIdea: Idea) => {
  set(ref(database, "ideas/" + ideaId), updatedIdea);
};

const toCamelCase = (sentenceCase: string) => {
  var out = "";
  sentenceCase.split(" ").forEach((el, idx) => {
    var add = el.toLowerCase();
    out += idx === 0 ? add : add[0].toUpperCase() + add.slice(1);
  });
  return out;
};

export {
  getEmployeeById,
  getEmployeeData,
  getIdeas,
  registerEmployee,
  registerIdea,
  updateIdea,
};
