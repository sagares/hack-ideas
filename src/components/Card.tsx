import React, { useContext } from "react";
import HackIdeasContext from "../context";
import { CardProps } from "../HackIdea.types";
import { updateIdea } from "../utils/ServiceUtil";

import "./card.less";

const Card = ({ id }: CardProps) => {
  const { currentUser, ideas } = useContext(HackIdeasContext);

  const idea = ideas[id];

  const isUpvoted = () => {
    return (
      idea.upvotes?.length > 0 &&
      idea.upvotes.some((upvoted) => upvoted.empId === currentUser.empId)
    );
  };

  const handleUpvoteClick = () => {
    const upvotes = idea.upvotes ? [...idea.upvotes] : [];
    if (!isUpvoted()) {
      upvotes.push(currentUser);
    } else {
      const index = upvotes.findIndex(
        (upvoted) => upvoted.empId === currentUser.empId
      );
      upvotes.splice(index, 1);
    }

    updateIdea(idea.id, {
      ...idea,
      upvotes,
    });
  };

  return (
    <div className="hack-idea--card card">
      <div className="card-content">
        <span className="card-title">{idea.name}</span>
        <p>{idea.description}</p>
        <div className="card-chips">
          {idea.tags.map((tag) => (
            <div className="chip" key={tag}>
              {tag}
            </div>
          ))}
        </div>
      </div>
      <div className="card-action">
        <a
          className={`card-action--button ${isUpvoted() ? "upvoted" : ""} ${
            idea.submittedBy === currentUser.empId ? "disabled" : ""
          }`}
          onClick={handleUpvoteClick}
        >
          <i className="material-icons">thumb_up</i>
          {idea.upvotes?.length > 0 && (
            <span className="upvotes-count">({idea.upvotes.length})</span>
          )}
        </a>
      </div>
    </div>
  );
};

export default Card;
