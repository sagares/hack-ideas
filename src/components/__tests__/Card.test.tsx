/**
 * @jest-environment jsdom
 */
import React from "react";
import { act, fireEvent, render } from "@testing-library/react";
import { updateIdea } from "../../utils/ServiceUtil";
import HackIdeasContext from "../../context";
import Card from "../Card";

jest.mock("../../utils/ServiceUtil");

describe("Card Component", () => {
  let element, component, setIdeas;

  beforeEach(() => {
    (updateIdea as jest.Mock).mockImplementation(() => {});

    setIdeas = jest.fn().mockImplementation(() => {});
    element = document.createElement("div");

    component = render(
      <HackIdeasContext.Provider
        value={{
          currentUser: { empId: "123456" },
          ideas: {
            idea1: {
              name: "idea 1",
              description: "idea 1",
              id: "idea1",
              tags: ["tag1", "tag2"],
              submittedBy: "123456",
              submittedOn: 100001,
            },
            idea2: {
              name: "idea 2",
              description: "idea 2",
              id: "idea1",
              tags: ["tag1", "tag2"],
              submittedBy: "123457",
              submittedOn: 100002,
              upvotes: [{ empId: "123456" }],
            },
          },
          setIdeas,
        }}
      >
        <Card id="idea1" />
      </HackIdeasContext.Provider>,
      {
        container: document.body.appendChild(element),
      }
    );
  });

  afterEach(() => {
    document.body.removeChild(element);
  });

  test("renders", () => {
    expect(component.container).toMatchSnapshot();
  });

  test("disables upvote button if card is submitted by current user", () => {
    expect(
      component.container
        .querySelector(".card-action--button")
        .classList.contains("disabled")
    ).toBe(true);
  });

  test("does not render upvotes count if no upvotes on idea", () => {
    expect(
      component.container.querySelector(".card-action--button .upvotes-count")
    ).toBe(null);
  });

  test("does not render upvotes count if no upvotes on idea", () => {
    component = render(
      <HackIdeasContext.Provider
        value={{
          currentUser: { empId: "123456" },
          ideas: {
            idea1: {
              name: "idea 1",
              description: "idea 1",
              id: "idea1",
              tags: ["tag1", "tag2"],
              submittedBy: "123456",
              submittedOn: 100001,
              upvotes: [],
            },
            idea2: {
              name: "idea 2",
              description: "idea 2",
              id: "idea1",
              tags: ["tag1", "tag2"],
              submittedBy: "123457",
              submittedOn: 100002,
              upvotes: [{ empId: "123456" }],
            },
          },
          setIdeas,
        }}
      >
        <Card id="idea1" />
      </HackIdeasContext.Provider>,
      {
        container: document.body.appendChild(element),
      }
    );
    expect(
      component.container.querySelector(".card-action--button .upvotes-count")
    ).toBe(null);
  });

  describe("Other's Idea", () => {
    let component;
    beforeEach(() => {
      component = render(
        <HackIdeasContext.Provider
          value={{
            currentUser: { empId: "123456" },
            ideas: {
              idea1: {
                name: "idea 1",
                description: "idea 1",
                id: "idea1",
                tags: ["tag1", "tag2"],
                submittedBy: "123456",
                submittedOn: 100001,
                upvotes: [],
              },
              idea2: {
                name: "idea 2",
                description: "idea 2",
                id: "idea2",
                tags: ["tag1", "tag2"],
                submittedBy: "123457",
                submittedOn: 100002,
                upvotes: [{ empId: "123456" }, { empId: "123458" }],
              },
            },
            setIdeas,
          }}
        >
          <Card id="idea2" />
        </HackIdeasContext.Provider>,
        {
          container: document.body.appendChild(element),
        }
      );
    });

    test("renders upvotes count", () => {
      expect(
        component.container.querySelector(".card-action--button .upvotes-count")
          .innerHTML
      ).toBe("(2)");
    });

    test("marks upvoted if current user has already upvoted the idea", () => {
      expect(
        component.container
          .querySelector(".card-action--button")
          .classList.contains("upvoted")
      ).toBe(true);
    });

    test("updates idea on click without current user inside upvotes if already upvoted", () => {
      const upvoteBtn = component.container.querySelector(
        ".card-action--button"
      );

      act(() => {
        fireEvent.click(upvoteBtn);
      });

      expect(updateIdea).toHaveBeenCalledWith("idea2", {
        description: "idea 2",
        id: "idea2",
        name: "idea 2",
        submittedBy: "123457",
        submittedOn: 100002,
        tags: ["tag1", "tag2"],
        upvotes: [{ empId: "123458" }],
      });
    });

    test("updates idea on click with current user inside upvotes if already not upvoted", () => {
      component = render(
        <HackIdeasContext.Provider
          value={{
            currentUser: { empId: "123456" },
            ideas: {
              idea1: {
                name: "idea 1",
                description: "idea 1",
                id: "idea1",
                tags: ["tag1", "tag2"],
                submittedBy: "123456",
                submittedOn: 100001,
                upvotes: [],
              },
              idea2: {
                name: "idea 2",
                description: "idea 2",
                id: "idea2",
                tags: ["tag1", "tag2"],
                submittedBy: "123457",
                submittedOn: 100002,
              },
            },
            setIdeas,
          }}
        >
          <Card id="idea2" />
        </HackIdeasContext.Provider>,
        {
          container: document.body.appendChild(element),
        }
      );

      const upvoteBtn = component.container.querySelector(
        ".card-action--button"
      );

      act(() => {
        fireEvent.click(upvoteBtn);
      });

      expect(updateIdea).toHaveBeenCalledWith("idea2", {
        description: "idea 2",
        id: "idea2",
        name: "idea 2",
        submittedBy: "123457",
        submittedOn: 100002,
        tags: ["tag1", "tag2"],
        upvotes: [{ empId: "123456" }],
      });
    });
  });
});
