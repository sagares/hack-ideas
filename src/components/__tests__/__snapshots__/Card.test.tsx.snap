// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component renders 1`] = `
<div>
  <div
    class="hack-idea--card card"
  >
    <div
      class="card-content"
    >
      <span
        class="card-title"
      >
        idea 1
      </span>
      <p>
        idea 1
      </p>
      <div
        class="card-chips"
      >
        <div
          class="chip"
        >
          tag1
        </div>
        <div
          class="chip"
        >
          tag2
        </div>
      </div>
    </div>
    <div
      class="card-action"
    >
      <a
        class="card-action--button  disabled"
      >
        <i
          class="material-icons"
        >
          thumb_up
        </i>
      </a>
    </div>
  </div>
</div>
`;
